plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,65 +4,318 @@
+
+instance Ord Rational where
+  compare (Rational n d) (Rational n' d') = compare (n * d') (n' * d)
+  Rational n d <= Rational n' d' = (n * d') <= (n' * d)
+  Rational n d >= Rational n' d' = (n * d') >= (n' * d)
+  Rational n d < Rational n' d' = (n * d') < (n' * d)
+  Rational n d > Rational n' d' = (n * d') > (n' * d)
+
+instance P.AdditiveSemigroup Rational where
+  {-# INLINEABLE (+) #-}
+  Rational n d + Rational n' d' =
+    let newNum = (n P.* d') P.+ (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.AdditiveMonoid Rational where
+  {-# INLINEABLE zero #-}
+  zero = Rational P.zero P.one
+
+instance P.AdditiveGroup Rational where
+  {-# INLINEABLE (-) #-}
+  Rational n d - Rational n' d' =
+    let newNum = (n P.* d') P.- (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeSemigroup Rational where
+  {-# INLINEABLE (*) #-}
+  Rational n d * Rational n' d' =
+    let newNum = n P.* n'
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeMonoid Rational where
+  {-# INLINEABLE one #-}
+  one = Rational P.one P.one
+
+instance P.ToData Rational where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (Rational n d) = P.toBuiltinData (n, d)
+
+instance P.FromData Rational where
+  {-# INLINEABLE fromBuiltinData #-}
+  fromBuiltinData dat = case P.fromBuiltinData dat of
+    P.Nothing -> P.Nothing
+    P.Just (n, d) -> if d P.== P.zero
+                     then Builtins.error ()
+                     else P.Just (n % d)
+
+instance P.UnsafeFromData Rational where
+  {-# INLINEABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData dat = case P.unsafeFromBuiltinData dat of
+    (n, d) -> if d P.== P.zero
+              then Builtins.error ()
+              else n % d
+
+instance ToJSON Rational where
+  toJSON (Rational n d) =
+    object
+      [ ("numerator", toJSON n)
+      , ("denominator", toJSON d)
+      ]
+
+instance FromJSON Rational where
+  parseJSON = withObject "Rational" Haskell.$ \obj -> do
+    n <- obj .: "numerator"
+    d <- obj .: "denominator"
+    if d Haskell.== 0
+      then Haskell.fail "Zero denominator is invalid."
+      else Haskell.pure (n % d)
+
+-- | Makes a 'Rational' from a numerator and a denominator.
+--
+-- = Important note
+--
+-- If given a zero denominator, this function will error. If you don't mind a
+-- size increase, and care about safety, use 'ratio' instead.
+{-# INLINEABLE (%) #-}
+(%) :: Integer -> Integer -> Rational
+n % d
+  | d P.== P.zero = Builtins.error ()
+  | d P.< P.zero = P.negate n % P.negate d
```

Comment:
Except I am not doing that? The constructor is deliberately left _un_exposed so that the invariants can be maintained without any risk of breakage. We even ensure this at serialization boundaries.

Reply:
Right, I can't read. Please document the invariant clearly, then.